make all classes haxe3 compatible

grepsuzette opened the pull request:
No idea if the team would find it worthwhile to have these new crypto classes also available for Haxe 3 (see https://github.com/HaxeFoundation/crypto/issues/6), in any case I wanted to post this simple PR.

grepsuzette committed:
make all classes haxe3 compatible

Simn commented:
Is there any point to this given that the Haxe 3 standard library has these classes anyway?

Well, I suppose it's nice to be able to do `-lib crypto` and have it just working with both Haxe 3 and 4. So I'm fine with making these changes.

grepsuzette commented:
> Is there any point to this given that the Haxe 3 standard library has these classes anyway?

I am under the impression Haxe 3.4.7 haxe.crypto only features:

* Adler32
* Base64
* BaseCode
* Crc32
* Hmac
* Md5
* Sha1
* Sha224
* Sha256

Thus missing:  Aes, BCrypt, BlowFish, Des, Pbkdf2, Ripemd160, Sha384, Sha512, TripleDes, TwoFish.
I'll let you check if that's actually the case cause I'm not sure how to be sure, but for example Aes has very nice possible applications and Sha512 begins to become used in some area (e.g. signing on some cryptocurrency exchange APIs). Arguably all of them will have nice use cases.

Gama11 commented:
`#if haxe3 @:extern #else extern #end` is quite silly / ugly. At that point you should just use `@:extern` in both cases, it still works in Haxe 4.

Gama11 commented:
Besides, it doesn't actually do anything, `haxe3` is still defined in Haxe 4 since it just means >= 3. ;)

grepsuzette committed:
fix silly conditional to use @:extern on all versions, added a paragraph in README to provide brief explanation for haxe 3 users

Gama11 commented on src/haxe/crypto/Sha224.hx:
```diff
@@ -27,9 +27,9 @@ import haxe.ds.Vector;
+    public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
         var sh = new Sha224();
-		var data = haxe.io.Bytes.ofString(s, encoding);
+		var data = haxe.io.Bytes.ofString(s #if !haxe3 , encoding #end );
```
I still don't think this works, given that `haxe3` is defined for both Haxe 3 and 4? `#if haxe4` should be better?

grepsuzette replied:
Yes indeed. Thanks

grepsuzette commented on src/haxe/crypto/Sha224.hx:
```diff
@@ -27,9 +27,9 @@ import haxe.ds.Vector;
 class Sha224 {
 
-    public static function encode( s:String, ?encoding : haxe.io.Encoding ) : String {
+    public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
```
```suggestion
    public static function encode( s:String #if haxe4 , ?encoding : haxe.io.Encoding #end ) : String {
```

grepsuzette commented on src/haxe/crypto/Sha224.hx:
```diff
@@ -27,9 +27,9 @@ import haxe.ds.Vector;
+    public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
         var sh = new Sha224();
-		var data = haxe.io.Bytes.ofString(s, encoding);
+		var data = haxe.io.Bytes.ofString(s #if !haxe3 , encoding #end );
```
```suggestion
		var data = haxe.io.Bytes.ofString(s #if haxe4 , encoding #end );
```

grepsuzette commented on src/haxe/crypto/Sha256.hx:
```diff
@@ -28,9 +28,9 @@ import haxe.ds.Vector;
 class Sha256 {
 
-	public static function encode( s:String, ?encoding : haxe.io.Encoding ) : String {
+	public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
```
```suggestion
	public static function encode( s:String #if haxe4 , ?encoding : haxe.io.Encoding #end ) : String {
```

grepsuzette commented on src/haxe/crypto/Sha256.hx:
```diff
@@ -28,9 +28,9 @@ import haxe.ds.Vector;
+	public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
 		var sh = new Sha256();
-		var data = haxe.io.Bytes.ofString(s, encoding);
+		var data = haxe.io.Bytes.ofString(s #if !haxe3 , encoding #end );
```
```suggestion
		var data = haxe.io.Bytes.ofString(s #if haxe4 , encoding #end );
```

grepsuzette commented on src/haxe/crypto/Sha384.hx:
```diff
@@ -3,7 +3,7 @@ package haxe.crypto;
 import haxe.Int64;
 import haxe.io.Bytes;
-import haxe.io.Encoding;
+#if !haxe3 import haxe.io.Encoding; #end
```
```suggestion
#if haxe4 import haxe.io.Encoding; #end
```

grepsuzette commented on src/haxe/crypto/Sha384.hx:
```diff
@@ -12,8 +12,8 @@ class Sha384
 
-    public static function encode( s:String, ?encoding : haxe.io.Encoding ) : String {
-		var data = haxe.io.Bytes.ofString(s, encoding);
+    public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
```
```suggestion
    public static function encode( s:String #if haxe4 , ?encoding : haxe.io.Encoding #end ) : String {
```

grepsuzette commented on src/haxe/crypto/Sha384.hx:
```diff
@@ -12,8 +12,8 @@ class Sha384
-    public static function encode( s:String, ?encoding : haxe.io.Encoding ) : String {
-		var data = haxe.io.Bytes.ofString(s, encoding);
+    public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
+		var data = haxe.io.Bytes.ofString(s #if !haxe3 , encoding #end );
```
```suggestion
		var data = haxe.io.Bytes.ofString(s #if haxe4 , encoding #end );
```

grepsuzette commented on src/haxe/crypto/Sha512.hx:
```diff
@@ -3,7 +3,7 @@ package haxe.crypto;
 import haxe.Int64;
 import haxe.io.Bytes;
-import haxe.io.Encoding;
+#if !haxe3 import haxe.io.Encoding; #end
```
```suggestion
#if haxe4 import haxe.io.Encoding; #end
```

grepsuzette commented on src/haxe/crypto/Sha512.hx:
```diff
@@ -12,8 +12,8 @@ class Sha512
 
-	public static function encode( s:String, ?encoding : haxe.io.Encoding ) : String {
-		var data = haxe.io.Bytes.ofString(s, encoding);
+	public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
```
```suggestion
	public static function encode( s:String #if haxe4 , ?encoding : haxe.io.Encoding #end ) : String {
```

grepsuzette commented on src/haxe/crypto/Sha512.hx:
```diff
@@ -12,8 +12,8 @@ class Sha512
-	public static function encode( s:String, ?encoding : haxe.io.Encoding ) : String {
-		var data = haxe.io.Bytes.ofString(s, encoding);
+	public static function encode( s:String #if !haxe3 , ?encoding : haxe.io.Encoding #end ) : String {
+		var data = haxe.io.Bytes.ofString(s #if !haxe3 , encoding #end );
```
```suggestion
		var data = haxe.io.Bytes.ofString(s #if haxe4 , encoding #end );
```

grepsuzette replied:
(Sorry seems like I should have made a batch, first time I use this resolve feature)

Commit:
Apply suggestions from code review

fix: #if haxe4 #end should be used since `#if haxe3` is also defined for haxe 4

flashultra commented:
Thank you for your pull request.  I just did some test using haxe 3.4..4 and everything works .
 Here is example for Sha512 :
  https://try.haxe.org/#B7266   and  http://try-haxe.mrcdk.com/#0b8f9
I think it's safe to merge the changes.

grepsuzette commented:
Maybe we should integrate automated unit tests for those, if not now at least soon afterwards.
